Improve linting logic to handle conditions that contain global references

nlfurniss opened the pull request:
Currently, `no-unguarded-globals` will pass all these lines:

```js
// Good
if (isBrowser() && window.location) { ... }

// Bad
if (window.location && isBrowser()) { ... }
if (isBrowser() || window.location)) { ... }
```

**This commit adds**:
* Logic to correctly handle global references in conditions
* Many test cases
* JSDoc comments

Thanks to @nickiaconis for his tests from #10

nlfurniss committed:
Improve linting logic to handle conditions that contain global references

scalvert commented on tests/lib/rules/no-unguarded-globals.js:
```diff
@@ -504,6 +504,72 @@ ruleTester.run('no-unguarded-globals', rule, {
+            }
+          }
+        });`,
+      parserOptions: {
```
You'll likely want to rebase and clean up these `parserOptions` usages now that the other PR has landed that removes multiple usages of them.

nlfurniss replied:
Done

scalvert commented on tests/lib/rules/no-unguarded-globals.js:
```diff
@@ -504,6 +504,72 @@ ruleTester.run('no-unguarded-globals', rule, {
+    {
+      code: `
+        import environment from 'ember-stdlib/utils/environment';
+        const { isBrowser } = environment
```
Can we add a test to ensure that aliases are still being captured and followed?

```js
const { isBrowser: isB } = environment; 
```

nlfurniss replied:
This doesn't work in master...

```js 
 1) no-unguarded-globals valid 
        import environment from 'ember-stdlib/utils/environment';
        const { isBrowser: isB } = environment;

        export default Ember.Component.extend({
          isInBrowser() {
            if (isB()) {
              return window.location;
            }
          }
        });:

      AssertionError [ERR_ASSERTION]: Should have no errors but had 1: [ { ruleId: 'no-unguarded-globals',...
```

scalvert replied:
OK, we should fix it. We have utilities that handle this in `ember-best-practices`, but because that repo was created after this one it's likely they weren't back-ported.

scalvert replied:
I lied, it doesn't look like we support aliases. We should put that in our TODOs.

scalvert commented on tests/lib/rules/no-unguarded-globals.js:
```diff
@@ -504,6 +504,72 @@ ruleTester.run('no-unguarded-globals', rule, {
+
+        export default Ember.Component.extend({
+          isInBrowser() {
+            if (environment.isBrowser() || window) {
```
Will this work if one does:

```js
return environment.isBrowser() || window;
```
?

nlfurniss replied:
That is marked as invalid. Have added a test for it.

scalvert replied:
Great, thanks!

scalvert reviewed (changes requested):
Thanks for adding these! I have another branch that's looking at expanding this rule to handle early returns and ternaries. It'd be great to let this one land and see where the holes are.

scalvert commented on lib/utils/bpr.js:
```diff
@@ -11,16 +16,34 @@ function _getConditions(node) {
 
   if (node.test.left) {
+    // If the global reference is part of the conditional and is on the left side (i.e. it's before the
+    // isBrowser check), return an empty array of callees to make isEnvironmentBrowser/isBrowser check false.
```
Can we refactor the conditional to remove the need for these comments? That may mean creating intermediate variables/functions to make the check more explicit. 

I'm much prefer this vs. comments as comments become out of date, but code doesn't.

nlfurniss replied:
Done

scalvert commented on lib/utils/bpr.js:
```diff
@@ -11,16 +16,34 @@ function _getConditions(node) {
+    // If the global reference is part of the conditional and is on the left side (i.e. it's before the
+    // isBrowser check), return an empty array of callees to make isEnvironmentBrowser/isBrowser check false.
+    if (isGlobalPartofCondition(node, globalReferenceNode) && node.test.left.object === globalReferenceNode) {
+      return [];
```
Can we create a constant above to represent this?

```js
const EMPTY_CALLEES = Object.freeze([]);
```

scalvert commented on lib/utils/bpr.js:
```diff
@@ -11,16 +16,34 @@ function _getConditions(node) {
 
   if (node.test.right) {
+    // If the global reference is part of the conditional and is on the right side (i.e. it's after the
+    // isBrowser check), and the operator is ||, return an empty array of callees to make
```
Same as above. If we can refactor this to remove the comment that would be ideal.

scalvert reviewed (changes requested):
Dropped a few more comments.

nlfurniss commented:
Updated after feedback. refactored the left/right side condition checking with separate functions.

Let me know if this fits the bill for what you had in mind

nlfurniss commented:
bumpitty bump bump @scalvert
